snapshot/volume-snapshotting-proposal.md

```diff
@@ -0,0 +1,412 @@
+The initial version is created as an external controller and provisioner (i.e. two separate binaries).
+
+* Goal 1: Enable *on-demand* snapshots of Kubernetes persistent volumes by application developers.
+
+    * Nongoal: Enable *automatic* periodic snapshotting for direct volumes in pods.
+
+* Goal 2: Expose standardized snapshotting operations to create, list and delete snapshots in Kubernetes REST API.
+
+* Goal 3: Implement snapshotting interface for Amazon EBS.
+
+* Goal 4: Implement snapshotting interface for GCE PDs.
+
+### Feature Roadmap
+
+Major features, planned for the first version:
+
+* On demand snapshots
+
+    * API to create new snapshots
+
+    * API to list snapshots available to the user
+
+    * API to delete existing snapshots
+
+    * API to create a persistent volume from a snapshot
+
+### Future Features
+
+The features that are not planned for the first version of the API bus should be considered in future versions:
+
+* Creating snapshots
+
+    * Scheduled and periodic snapshots
+
+    * Aplication initiated on-demand snapshot creation
+
+    * Support snapshot per PVC, pod or StatefulSet
+
+    * Support snapshots for non-cloud storage volumes (i.e. plugins that require actions to be triggered from the node)
+
+    * Support application-consistent snapshots (coordinate distributed snapshots across multiple volumes)
+
+    * Enable to create a pod/statefulsets with snapshots
+
+* List snapshots
+
+    * Enable to get the list of all snapshots for a specified persistent volume
+
+    * Enable to get the list of all snapshots for a pod/StatefulSet
+
+* Delete snapshots
+
+    * Enable to automatic garbage collect older snapshots when storage is limited
+
+* Quota management
+
+    * Enable to set quota for limiting how many snapshots could be taken and saved
+
+    * When quota is exceeded, delete the oldest snapshots automatically
+
+## Requirements
+
+### Performance
+
+* Time SLA from issuing a snapshot to completion:
+
+* The period we are interested is the time between the scheduled snapshot time and the time the snapshot is finishes uploading to its storage location.
+
+* This should be on the order of a few minutes.
+
+### Reliability
+
+* Data corruption
+
+    * Though it is generally recommended to stop application writes before executing the snapshot command, we will not do this automatically for several reasons:
+
+        * GCE and Amazon can create snapshots while the application is running.
+
+        * Stopping application writes cannot be done from the master and varies by application, so doing so will introduce unnecessary complexity and permission issues in the code.
+
+        * Most file systems and server applications are (and should be) able to restore inconsistent snapshots the same way as a disk that underwent an unclean shutdown.
+
+    * The data consistency would be best-effort only: e.g., call fsfreeze prior the snapshot on filesystems that support it.
+
```

Comment:
IMO, data consistency when performing snapshot should be left to the storage provider. As  a second thought, the volume plugins could report/declare a  field/attribute like `online`  or `offline` ( their snapshot capability) in one of the spec. So that the controller can be better taught to perform snapshot. This may come handy in future when we expand more use cases.